### PR TITLE
Make LocalFileSystem::put atomic (#3780)

### LINKED ISSUE #3780: LocalFileSystem::put is not Atomic
**Describe the bug**

`ObjectStore::put` should be an atomic operation, that is the file should only be visible to listing operations once it has been written in its entirety. For multipart uploads we upload to a staging path and move it into place, we should do the same for regular put operations.

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3780

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/local.rs [hunk 1]
```diff
@@ -553,9 +543,40 @@ impl ObjectStore for LocalFileSystem {
     }
 }
 
-fn get_upload_stage_path(dest: &std::path::Path, multipart_id: &MultipartId) -> PathBuf {
+/// Generates a unique file path `{base}#{suffix}`, returning the opened `File` and `suffix`
+///
+/// Creates any directories if necessary
+fn new_staged_upload(base: &PathBuf) -> Result<(File, String)> {
```
Comment (tustvold) [inline object_store/src/local.rs:549]: This combines the logic of `open_writeable_file` and `get_upload_stage_path`

**Review by crepererum — COMMENTED** [review]
Could you add a test validating that the temporary files aren't listed and that listing while temporary files are present doesn't error?

**Comment by crepererum** [comment]
I'm also wondering if we could be slightly smarter on Linux: instead of creating an ordinary file that will never get pruned if the write fails half-way, we could create an unlinked file (`open(2)` w/ `O_TMPFILE`) and link it once the write is done.

**Comment by tustvold** [comment]
> Could you add a test validating that the temporary files aren't listed and that listing while temporary files are present doesn't error?

There already is such a test - https://github.com/apache/arrow-rs/blob/master/object_store/src/lib.rs#L836

File: object_store/src/lib.rs [hunk 2]
```diff
@@ -840,6 +840,12 @@ mod tests {
             crate::Error::NotFound { .. }
         ));
 
+        let files = flatten_list_stream(storage, None).await.unwrap();
```
Comment (alamb) [inline object_store/src/lib.rs:843]: I didn't see a test for atomic `puts` (this appears to be testing atomic `put_multipart`). 

Maybe I am misreading the code, but I think there should be a separate test to ensure we don't accidentally break `put` in the future (and that the other implementations are also atomic)

**Review by alamb — COMMENTED** [review]
The code looks good to me (and I learned a few things 👍 )

I think another test is warranted (unless I am missing it) otherwise it looks (very) good

File: object_store/src/lib.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/lib.rs:843]: I wasn't really sure how to reliably test this as it depends on the latency of the write... I'll have a think

Comment (alamb) [reply object_store/src/lib.rs:843]: Maybe a stream where you can send one `bytes`, verify the file doesn't appear, and then close the stream?

Comment (tustvold) [reply object_store/src/lib.rs:843]: `put` doesn't take a stream, it only takes a `Bytes`...

Comment (alamb) [reply object_store/src/lib.rs:843]: Ah, well then I agree testing this is likely going to be hard and intrusive 👍

**Review by alamb — APPROVED** [review]

**Pull request merged**